tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithArrays.scala

```diff
@@ -0,0 +1,83 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt.implicitConversions.Crossable
+import at.forsyte.apalache.tla.bmcmt.rewriter.ConstSimplifierForSmt
+import at.forsyte.apalache.tla.bmcmt.types.{BoolT, CellT, FinSetT, PowSetT}
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, SymbState, SymbStateRewriter, types}
+import at.forsyte.apalache.tla.lir.TlaEx
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+
+class SetInRuleWithArrays(rewriter: SymbStateRewriter) extends SetInRule(rewriter) {
+  private val simplifier = new ConstSimplifierForSmt
+
+  override protected def powSetIn(state: SymbState, powsetCell: ArenaCell, elemCell: ArenaCell): SymbState = {
+    def checkType: PartialFunction[(CellT, CellT), Unit] = {
+      case (PowSetT(FinSetT(expectedType)), FinSetT(actualType)) =>
+        assert(expectedType == actualType)
+    }
+    // double check that the type finder did its job
+    checkType(powsetCell.cellType, elemCell.cellType)
+
+    val setElems = state.arena.getHas(elemCell)
+    val powsetDomain = state.arena.getDom(powsetCell)
+    val powsetDomainElems = state.arena.getHas(powsetDomain)
+    // EqConstraints need to be generated, since missing in-relations, e.g. in sets of tuples, will lead to errors.
+    // TODO: Inlining this method is pointless. We should consider handling tuples and other structures natively in SMT.
+    var newState = rewriter.lazyEq.cacheEqConstraints(state, setElems cross powsetDomainElems)
+
+    def isInPowset(setElem: ArenaCell): TlaEx = {
+      newState = newState.updateArena(_.appendCell(BoolT()))
+      val pred = newState.arena.topCell
+
+      def isInAndEqSetElem(powsetDomainElem: ArenaCell) = {
+        simplifier.simplifyShallow(tla.and(tla.in(powsetDomainElem.toNameEx, powsetDomain.toNameEx),
+                tla.eql(powsetDomainElem.toNameEx, setElem.toNameEx)))
+      }
+
+      val elemsInAndEqLeftElem = powsetDomainElems.map(isInAndEqSetElem)
+      rewriter.solverContext.assertGroundExpr(simplifier.simplifyShallow(tla.equiv(pred.toNameEx,
+                  tla.and(tla.in(setElem.toNameEx, elemCell.toNameEx), tla.or(elemsInAndEqLeftElem: _*)))))
```

Comment:
Why is it `and` and not an implication?

Reply:
If it's correct, a comment that describes the formula, e.g, as a TLA+ formula, would help.

Reply:
The original rule uses implication (see `LazyEquality.subsetEq`), but I decided to use conjunction here because I think there is no case in which `tla.in(setElem.toNameEx, elemCell.toNameEx)` evaluates to false. In the beginning of the method we explicitly have `setElems = state.arena.getHas(elemCell)`.

I'm not sure what a TLA+ formula of this would be, since this rule is applied for, e.g., {1} \in SUBSET {1,2}. Is this what you mean?

Reply:
Well, how about the result of `{1} \ {1}`? The arena would point to one cell (the removed `1`), but that cell would not be in the set (difference). I think you need an implication.

Reply:
> I'm not sure what a TLA+ formula of this would be, since this rule is applied for, e.g., {1} \in SUBSET {1,2}. Is this what you mean?

What I mean is that sometimes it is useful to write in comments the constraints that you are constructing. TLA+ is still more readable than the expression in `assertGroundExpr`.

Reply:
> Well, how about the result of `{1} \ {1}`? The arena would point to one cell (the removed `1`), but that cell would not be in the set (difference). I think you need an implication.

I believe in this case the arena would point to a set cell with no edges, i.e. the set cell does not point to any other cells, no? Before `powSetIn` is called, `apply` calls `rewriter.rewriteUntilDone` twice, passing the lhs and rhs arguments, so this should not be an issue if the other rules are working as intended. We would have `{1} \in SUBSET ({1} \ {1})` rewritten as `{1} \in SUBSET ({ })`, since the `SUBSET` expression is not evaluated, but its argument is.

I tried a simple example here and adding `{1} \in SUBSET ({1} \ {1})` as a constraint to the spec leads to a deadlock, while adding `{1} \in SUBSET ({1} \ {2})` is fine, as expected. If I'm missing something, let me know.

> What I mean is that sometimes it is useful to write in comments the constraints that you are constructing. TLA+ is still more readable than the expression in `assertGroundExpr`.

I do find the `assertGroundExpr` quite readable, but I see your point. I did not add comments regarding the constraints being constructed because I didn't see any on the existing rules and simply followed that.

Reply:
In this simple case, you are right, the expression `{1} \ {1}` will be probably simplified. But there are less certain cases. For instance, consider the expression `{x} \ {y}`, where `x` and `y` are state variables. Then the arena of the set will point to a single cell. However, the set is empty when `x = y`.

Reply:
I tried the example with state variable that you suggested and it still works as intended. The set minus operation needs to be evaluated at some point, and from what I saw in the smt constraints generated it is rewritten as set filter. If `{x} \ {y}` points to `x` in the case where `x = y` then there is a problem with how the arena is being updated, no?

Reply:
The arena is an overapproximation of the set contents. In case like `{x} \ {y}`, the rewriting rules cannot statically figure out that the set is empty in case of `x = y`. So the arena will still have one edge, even though the SMT model will have this edge evaluated to false. If you look at how set filter is implemented, it does not remove edges from the arena, but disables them in SMT when the predicate evaluates to false. If we could statically figure out which edges are disabled, we would not need set-membership predicates (or arrays) in SMT.

Reply:
I see. This does explain the behavior of filter, which I found peculiar. For some reason I got the impression that this computation was done by the rules, but as you said we would essentially be doing the solver's job in Apalache if it was the case. I will make the change from conjunction to implication.